drivers: Driver for the WM8960 codec.

robert-hh opened the pull request:
This PR adds to the SAI support of PR #8322. It is a driver for the MW8960 codec,
whis is assembled for the MIMXRT1xxx_DEV boards and is available
at  breakout boards as well. The driver is required to get any sound out of
the WM8960 codec.

Including the driver into the MicroPython repository allows to put it by
default into the frozen bytecode of MIMXRT_1xxx_DEV boards.

The driver itself has been tested as working with the MIMXRT boards
and a Sparkfun WM8960 breakout board. It implements the initialization,
basic methods and some enhanced methods like 3D, ALC or Soft Mute.

robert-hh committed:
wm8960: Driver for the WM8960 codec.

This codec is assembled for the MIMXRT1xxx_DEV boards and available
for WM8960 breakout boards as well.

The driver itself has been tested as working with the MIMXRT boards
and a Sparkfun WM6890 breakout board. It implements the initialization,
basic methods and some enhanced methods like 3D, ALC or Soft Mute.

dpgeorge commented on drivers/codec/wm8960.py:
```diff
@@ -0,0 +1,940 @@
+        0x0050, 0x0050, 0x0050, 0x0000, 0x0000, 0x0000, 0x0000,
+        0x0040, 0x0000, 0x0000, 0x0050, 0x0050, 0x0000, 0x0002,
+        0x0037, 0x004d, 0x0080, 0x0008, 0x0031, 0x0026, 0x00e9
+    ]
```
If possible, better to make this a tuple, then it has a chance to be optimised by the compiler.

robert-hh replied:
I'm writing back into that table. So a tuple is not feasible.

dpgeorge commented on drivers/codec/wm8960.py:
```diff
@@ -0,0 +1,940 @@
+        self.write_reg(_WM8960_ADDCTL4, 0x60)  # Set GPIO1 to 0.
+
+        self.write_reg(_WM8960_BYPASS1, 0x0)
+        self.write_reg(_WM8960_BYPASS2, 0x0)
```
Since Python is a high-level language, I can't help but think there might be ways to make this code simpler and not so long/verbose.

Eg, since a lot of these pairs of regs are one after the other, there could be a `write_reg_pair` method that writes 2 at once.

And you could make a small class to manage reg read/writes with more efficient notation like subscript.   Eg, schematically:
```python
class Regs:
    def __getitem__(self, reg):
        return self.cache[reg]
    def __setitem__(self, reg, value):
        self.cache[reg] = value
        i2c.writeto_mem(addr, reg, value)

...

# use it like this
r = self.regs
r[_WM8960_BYPASS1] = 0x0
```
```

robert-hh replied:
Since reg is a list, `self.read_reg(a)` could already now be written as `self.reg[a]`, but I kept it as a function for symmetry.

dpgeorge commented:
> Including the driver into the MicroPython repository allows to put it by
default into the frozen bytecode of MIMXRT_1xxx_DEV boards.

This driver could also maybe go in `micropython-lib`, and freezing it in firmware would still be possible.

robert-hh commented:
> This driver could also maybe go in micropython-lib, and freezing it in firmware would still be possible.

But that would assume that micropython-lib exists and is at a fixed relative position to the micropython tree.

robert-hh committed:
wm8960: Rework the code to a more pythonic style.

Following the suggestions of @dpgeorge:
- create a Regs class, which encapsulates accesses to the hardeware
  and holds the register cache.
- Change WM8960 register accesses into looking like array accesses.
- Combine modifications of the same register into one operation.
- Put assignments to L/R registers with the same value into one line.

Black had to be told to leave the code untouched, because it shifted
the whole WM8960 class.

robert-hh committed:
WM8960: Replace the register list by an array.

This array is initialized by a tuple with the power-on values.

robert-hh committed:
WM8960: Rewrite the method set_module() to a shorter style.

Which also made a few register shift definitions obsolete.

robert-hh committed:
WM8960: Change set_/get_volume() and editorial changes.

- remove unneded statement in set_volume()
- masked additional bits in get_volume()
- rewrote a line in __setitem__()
- shortened register access in three functions.

robert-hh commented:
@dpgeorge 

> This driver could also maybe go in micropython-lib, and freezing it in firmware would still be possible.

The code changes you suggested are done. About the `maybe`: should I read the as a strong suggestion? Then, does $(MPY_LIB_DIR) point in the firmware images build system to the micropython-lib directory?

robert-hh committed:
wm8960: Add a deemphasis filter method.

robert-hh committed:
wm8960: Simplify setting master mode and channel swap.

robert-hh committed:
WM8960: Drop the _WM8960 prefix and add register pair assignment.

The prefix drop seems to change about every line of code. So it looks
like a larger change than it is, but the code looks better.

robert-hh committed:
wm8960: Refactoring the codec.

- Normalize volume setting to the range 0-100.
- Use a table to simplify the code.
- Reformat the documentation.
- Add a default value for i2c_address.
- Some reformatting made by black.

dpgeorge commented:
> About the `maybe`: should I read the as a strong suggestion? Then, does $(MPY_LIB_DIR) point in the firmware images build system to the micropython-lib directory?

It would be good to put this in `micropython-lib`, but it's not ready yet to take such drivers, in particular it's not worked out how to make the documentation work (we may need to have separate docs for `micropython-lib`...).

So for now let's keep this driver in this repo.  There is anyway already the `lsm6dsox` driver, among others.

dpgeorge commented on drivers/codec/wm8960.py:
```diff
@@ -0,0 +1,783 @@
+
+# Config symbol names
+# Modules
+module_ADC = const(0)  # ADC module in WM8960
```
These constants, and all the ones below, should be FULL_CAPS to match other constants (eg `I2S.STEREO`).

robert-hh replied:
Done.

dpgeorge commented on drivers/codec/wm8960.py:
```diff
@@ -0,0 +1,783 @@
+sysclk_PLL = const(1)  # sysclk source from internal PLL
+
+
+# fmt: off
```
Please move this `fmt: off` down below to just before the `cache = ...` line.

robert-hh replied:
Done

dpgeorge commented on drivers/codec/wm8960.py:
```diff
@@ -0,0 +1,783 @@
+        return (int((self.regs[regnum] & vol_max) * 100 / vol_max + 0.5),
+                int((self.regs[regnum + 1] & vol_max) * 100 / vol_max + 0.5))
+
+    def volume(self, module, volume_l=None, volume_r=None):
```
This method is probably not needed?  Since it's just a shortcut for set/get.

robert-hh replied:
It prefer keeping the volume() method and drop set_volume() and get_volume().

robert-hh replied:
Done and pushed.

dpgeorge commented on drivers/codec/wm8960.py:
```diff
@@ -0,0 +1,783 @@
+        else:
+            raise ValueError("Invalid route")
+
+    def set_left_input(self, input):
```
It could be possible to simplify this method by having the input selection values more complex types.

Eg:
```python
INPUT_CLOSED = None
INPUT_MIC1 = (0x138, 0x117)
INPUT_MIC2 = (0x178, 0x117)
INPUT_MIC3 = (0x1b8, 0x117)
INPUT_LINE2 = (0, 0xe)
INPUT_LINE3 = (0, 0x70)

def set_left_input(self, input):
    if input is None:
        regs[_POWER1] = (_POWER1_AINL_MASK | _POWER1_ADCL_MASK, 0)
    elif input[0] == 0:
        regs[_POWER1] = (0, _POWER1_AINL_MASK | _POWER1_ADCL_MASK)
        regs[_INBMIX1] = input
    else:
        regs[_POWER1] = (0, _POWER1_AINL_MASK | _POWER1_ADCL_MASK | _POWER1_MICB_MASK)
        regs[_LINPATH] = input[0]
        regs[_LINVOL] = input[1]
```
And `set_right_input()` would also be simplified, and can use the same constants.

robert-hh replied:
That looks good, but I would like to do it in the same way as for the volumes, with an internal dictionary of the value pairs. Then INPUT_XXXX being method arguments could stay an integer constants, and the value can be checked for a valid value.

robert-hh replied:
Done and pushed.

dpgeorge commented on docs/reference/wm8960.rst:
```diff
@@ -0,0 +1,420 @@
+.. _wm8960:
```
please move this to `docs/library/wm8960.rst`

robert-hh replied:
Done.

robert-hh committed:
wm8960: Set all const names to FULL_CAPS.

robert-hh committed:
wm8960: Drop set_volume() and get_volume() in favour of volume().

robert-hh committed:
wm8960: Simplify set_left_input() and set_right_input().

According to the suggestion of @dpgeorge.
Fix a few omissions of the previous commits as well.

robert-hh committed:
wm8960: Reformatted with black and removed a reference in docs.

Still there is a warning about the wm8960.rst document not contained
in any doc tree. What would be the best place?

robert-hh committed:
wm8960: Add a paragraph in library/index.rst for drivers.

Silencing the warning about a missing toctree entry.

robert-hh committed:
wm8960: Set the DAC on init() to the reset values.

For some reason, this is required for one of my boards, and does not
hurt on the other ones.

dpgeorge commented:
Squashed and merged in 56b331ace6b7f01bacdb8b5c66f0f517670abfdc, with some reformatting of the docs.
